tests/test_semaphore_tracker.py

```diff
@@ -0,0 +1,210 @@
+    @pytest.mark.skipif(sys.platform == "win32",
+                        reason="no semaphore_tracker on windows")
+    def test_semaphore_tracker(self):
+        #
+        # Check that killing process does not leak named semaphores
+        #
+        import subprocess
+        cmd = '''if 1:
+            import time, os
+            from loky.backend.synchronize import Lock
+
+            # close manually the read end of the pipe in the child process
+            # because pass_fds does not exist for python < 3.2
+            os.close(%d)
+
+            lock1 = Lock()
+            lock2 = Lock()
+            os.write(%d, lock1._semlock.name.encode("ascii") + b"\\n")
+            os.write(%d, lock2._semlock.name.encode("ascii") + b"\\n")
+            time.sleep(10)
+        '''
+        r, w = os.pipe()
+
+        if sys.version_info[:2] >= (3, 2):
+            fd_kws = {'pass_fds': [w, r]}
+        else:
+            fd_kws = {'close_fds': False}
+        p = subprocess.Popen([sys.executable,
+                             '-E', '-c', cmd % (r, w, w)],
+                             stderr=subprocess.PIPE,
+                             **fd_kws)
+        os.close(w)
+        with io.open(r, 'rb', closefd=True) as f:
+            name1 = f.readline().rstrip().decode('ascii')
+            name2 = f.readline().rstrip().decode('ascii')
+
+        # subprocess holding a reference to lock1 is still alive, so this call
+        # should succeed
+        sem_unlink(name1)
+        p.terminate()
+        p.wait()
+        time.sleep(2.0)
+        with pytest.raises(OSError) as ctx:
+            sem_unlink(name2)
+        # docs say it should be ENOENT, but OSX seems to give EINVAL
+        assert ctx.value.errno in (errno.ENOENT, errno.EINVAL)
+        err = p.stderr.read().decode('utf-8')
+        p.stderr.close()
+        expected = ('semaphore_tracker: There appear to be 2 leaked '
+                    'semaphores')
+        assert re.search(expected, err) is not None
+
+        # lock1 is still registered, but was destroyed externally: the tracker
+        # is expected to complain.
+        expected = ("semaphore_tracker: %s: (OSError\\(%d|"
+                    "FileNotFoundError)" % (name1, errno.ENOENT))
+        assert re.search(expected, err) is not None
+
+    def check_semaphore_tracker_death(self, signum, should_die):
+        # bpo-31310: if the semaphore tracker process has died, it should
+        # be restarted implicitly.
+        from loky.backend.semaphore_tracker import _semaphore_tracker
+        # _semaphore_tracker.ensure_running()
+        pid = _semaphore_tracker._pid
+        if pid is not None:
+            os.kill(pid, signal.SIGKILL)
+            os.waitpid(pid, 0)
+        with warnings.catch_warnings():
+            warnings.simplefilter("ignore")
+            _semaphore_tracker.ensure_running()
+            # in python < 3.3 , the race condition described in bpo-33613 still
```

Comment:
I backported a fix for a race condition from `multiprocessing`, but it works only for python versions > 3.3. This race condition is subtle enough so that it should happen rarely, and is fairly specific. @ogrisel are you find with leaving this as is?

Reply:
or we could skip that tests under Python 2.7. As you prefer. I don't care about Python 2 anymore :)